opm/simulators/linalg/bda/rocsparseSolverBackend.cpp

```diff
@@ -0,0 +1,614 @@
+// MultisegmentWellContribution includes the cuda runtime if found by CMake
+// this leads to inclusion of both amd_hip_vector_types.h and vector_types.h
+// which both define vector types like uchar2, short3 and double4.
+#undef HAVE_CUDA
+
+#include <opm/simulators/linalg/bda/rocsparseSolverBackend.hpp>
+
+#include <opm/simulators/linalg/bda/BdaResult.hpp>
+
+#include <hip/hip_runtime_api.h>
+#include <hip/hip_version.h>
+
+#define HIP_CHECK(stat)                               \
+    {                                                 \
+        if(stat != hipSuccess)                        \
+        {                                             \
+            OPM_THROW(std::logic_error, "HIP error"); \
+        }                                             \
+    }
+
+#define ROCSPARSE_CHECK(stat)                               \
+    {                                                       \
+        if(stat != rocsparse_status_success)                \
+        {                                                   \
+            OPM_THROW(std::logic_error, "rocsparse error"); \
+        }                                                   \
+    }
+
+#define ROCBLAS_CHECK(stat)                               \
+    {                                                     \
+        if(stat != rocblas_status_success)                \
+        {                                                 \
+            OPM_THROW(std::logic_error, "rocblas error"); \
+        }                                                 \
+    }
+
+
+namespace Opm
+{
+namespace Accelerator
+{
+
+using Opm::OpmLog;
+using Dune::Timer;
+
+template <unsigned int block_size>
+rocsparseSolverBackend<block_size>::rocsparseSolverBackend(int verbosity_, int maxit_, double tolerance_, unsigned int platformID_, unsigned int deviceID_) : BdaSolver<block_size>(verbosity_, maxit_, tolerance_, platformID_, deviceID_) {
+    hipDevice_t device;
+    if(hipDeviceGet(&device, deviceID) != hipSuccess)
+    {
+        OPM_THROW(std::logic_error, "HIP Error: could not get device");
+    }
+
+    ROCSPARSE_CHECK(rocsparse_create_handle(&handle));
+    ROCBLAS_CHECK(rocblas_create_handle(&blas_handle));
+
+    ROCSPARSE_CHECK(rocsparse_get_version(handle, &ver));
+    ROCSPARSE_CHECK(rocsparse_get_git_rev(handle, rev));
+
+    std::ostringstream out;
+    out << "rocSPARSE version: " << ver / 100000 << "." << ver / 100 % 1000 << "."
+        << ver % 100 << "-" << rev << "\n";
+    OpmLog::info(out.str());
+
+    HIP_CHECK(hipStreamCreate(&stream));
+    ROCSPARSE_CHECK(rocsparse_set_stream(handle, stream));
+    ROCBLAS_CHECK(rocblas_set_stream(blas_handle, stream));
+}
+
+
+template <unsigned int block_size>
+rocsparseSolverBackend<block_size>::~rocsparseSolverBackend() {
+    HIP_CHECK(hipStreamSynchronize(stream));
+    HIP_CHECK(hipStreamDestroy(stream));
+    ROCSPARSE_CHECK(rocsparse_destroy_handle(handle));
+    ROCBLAS_CHECK(rocblas_destroy_handle(blas_handle));
```

Comment:
Since the `_CHECK` macros potentially call `OPM_THROW`, this is a destructor that will (potentially) throw an exception.  We should really try to avoid that.  Is there no other way of reporting failures here?

Reply:
We could write out the macro (or write a new one), replacing the `OPM_THROW` with something like `OpmLog::error()`.

The function `rocsparse_destroy_handle()` can perform some throwing and catching ([here](https://github.com/ROCmSoftwarePlatform/rocSPARSE/blob/82e7d8f09201ad55d08e62199b8ccc50251b1423/library/src/rocsparse_auxiliary.cpp#L79) and [here](https://github.com/ROCmSoftwarePlatform/rocSPARSE/blob/ba673954329423de932cc9b8d8365c0c277a4062/library/src/include/utility.h#L248)), but always returns a `rocsparse_status`.

Reply:
Changed to
```
    hipError_t hipstatus = hipStreamSynchronize(stream);
    if(hipstatus != hipSuccess){
        OpmLog::error("Could not synchronize with hipStream");
    }
```
could also write as
```
    if(hipStreamSynchronize(stream) != hipSuccess){
        OpmLog::error("Could not synchronize with hipStream");
    }
```